### PR TITLE
Performance optimizations for parquet sub-rowgroup reader.

### PR DESCRIPTION
**Pull request by nvdbaranec**
This PR implements a basket of optimizations for the parquet reader to bring non-chunked reads close  to par following the merge of the sub-rowgroup reader.  

The primary culprit for the performance hit was that in the case where we perform no splits, we were making a full copy of all of the pages into the subpass struct (including a pinned memory allocation).  This is unnecessary because we can just represent the pages in the subpass as a span that wraps the existing pages in the pass. 

In addition, several `hostdevice_vector`s used for work that could be done entirely device-side were converted to `rmm::device_uvector`.

Finally, I converted a number of functions that were taking hostdevice_vectors to use spans instead and added some missing operators to the `hostdevice_vector` class itself.

This PR doesn't recover all the time (there is some new work that we have to do in all cases) but it takes out most of the sting. A sample of some of the benchmarks that were most notably affected:

```
                       Original Time      Sub-rowgroup-implementation       This PR
parquet_read_decode
Int, device buffer 0   29260860778        26373181343                       28121328587
Int, device buffer 1   30692134492        27474241282                       29495189226

parquet_read_chunks
Int, device buffer     33895028252        29986276949                       32293548191
Float, device buffer   57055985251        49640274260                       55795392897
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/page_hdr.cu [hunk 1]
```diff
@@ -396,7 +396,7 @@ void __launch_bounds__(128) gpuDecodePageHeaders(ColumnChunkDesc* chunks,
     }
     num_values    = bs->ck.num_values;
     page_info     = chunk_pages ? chunk_pages[chunk].pages : nullptr;
-    max_num_pages = page_info ? bs->ck.max_num_pages : 0;
+    max_num_pages = page_info ? (bs->ck.num_data_pages + bs->ck.num_dict_pages) : 0;
```
Comment (ttnghia) [inline cpp/src/io/parquet/page_hdr.cu:399]: Why do we have this?

Comment (nvdbaranec) [reply cpp/src/io/parquet/page_hdr.cu:399]: `max_num_pages` was being computed on the host and stored in the chunk struct, but it was only ever used here, and we already have everything we need to compute it.  So I just removed it from the struct and compute it on the spot.

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 2]
```diff
@@ -549,8 +551,64 @@ struct get_page_span {
   }
 };
 
+/**
+ * @brief Return the span of page indices for a given column index
+ *
```
Comment (ttnghia) [inline cpp/src/io/parquet/reader_impl_chunking.cu:556]:
```suggestion
```

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 3]
```diff
@@ -549,8 +551,64 @@ struct get_page_span {
...
+  CUDF_HOST_DEVICE size_t operator()(page_span const& s) const { return s.end - s.start; }
+};
+
+/**
+ * @brief Return the size of a span in an array of spans, handling out-of-bounds indices.
+ *
+ */
+struct get_span_size_by_index {
+  cudf::device_span<page_span const> page_indices;
+  __device__ size_t operator()(size_t i) const
```
Comment (ttnghia) [inline cpp/src/io/parquet/reader_impl_chunking.cu:581]:
```suggestion
  cudf::device_span<page_span const> page_indices;
  
  __device__ size_t operator()(size_t i) const
```

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 4]
```diff
@@ -549,8 +551,64 @@ struct get_page_span {
...
+  {
+    return i >= page_indices.size() ? 0 : page_indices[i].end - page_indices[i].start;
+  }
+};
+
+/**
+ * @brief Copy page from appropriate source location (as defined by page_offsets) to the destination
+ * location, and store the index mapping.
+ *
```
Comment (ttnghia) [inline cpp/src/io/parquet/reader_impl_chunking.cu:590]:
```suggestion
```

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 5]
```diff
@@ -1319,37 +1382,50 @@ void reader::impl::setup_next_subpass(bool uses_custom_row_bounds)
...
-                     subpass.page_src_index.begin() + page_count,
-                     subpass.page_src_index.begin() + page_count + num_column_pages,
-                     page_indices[c_idx].start);
-    page_count += num_column_pages;
+  // check to see if we are processing the entire pass (enabling us to skip a bunch of work)
+  subpass.single_subpass = total_pages == pass.pages.size();
+
+  // in the single pass case, no page copying is necessary - just use what's in the pass itself
+  if (subpass.single_subpass) {
+    subpass.pages = pass.pages;
```
Comment (ttnghia) [inline cpp/src/io/parquet/reader_impl_chunking.cu:1390]: Does it need `std::move`?

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 6]
```diff
@@ -368,6 +371,71 @@ std::string encoding_to_string(Encoding encoding)
...
+  //
+  // 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3
+  //
+  // We also need to preserve key-relative page ordering, so we need to use a stable sort.
+  rmm::device_uvector<int32_t> page_keys{unsorted_pages.size(), stream};
+  thrust::transform(rmm::exec_policy_nosync(stream),
+                    unsorted_pages.begin(),
+                    unsorted_pages.end(),
+                    page_keys.begin(),
+                    [chunks = chunks.begin()] __device__(PageInfo const& page) {
```
Comment (ttnghia) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:414]: How can you compile this without `proclaim_return_type`?

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 7]
```diff
@@ -368,6 +371,71 @@ std::string encoding_to_string(Encoding encoding)
...
+                             sort_indices.begin(),
+                             thrust::less<int>());
+  auto pass_pages =
+    cudf::detail::hostdevice_vector<PageInfo>(unsorted_pages.size(), unsorted_pages.size(), stream);
+  thrust::transform(
+    rmm::exec_policy_nosync(stream),
+    sort_indices.begin(),
+    sort_indices.end(),
+    pass_pages.d_begin(),
+    [unsorted_pages = unsorted_pages.begin()] __device__(int32_t i) { return unsorted_pages[i]; });
```
Comment (ttnghia) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:434]: `proclaim_return_type`?

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 8]
```diff
@@ -377,21 +445,33 @@ void decode_page_headers(pass_intermediate_data& pass,
...
+    thrust::plus<size_t>{});
+  rmm::device_uvector<chunk_page_info> d_chunk_page_info(pass.chunks.size(), stream);
+  thrust::for_each(rmm::exec_policy_nosync(stream),
+                   iter,
+                   iter + pass.chunks.size(),
+                   [cpi               = d_chunk_page_info.begin(),
+                    chunk_page_counts = chunk_page_counts.begin(),
+                    unsorted_pages    = unsorted_pages.begin()] __device__(size_t i) {
+                     cpi[i].pages = &unsorted_pages[chunk_page_counts[i]];
+                   });
```
Comment (ttnghia) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:470]: `proclaim_return_type`?

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 9]
```diff
@@ -1319,37 +1382,50 @@ void reader::impl::setup_next_subpass(bool uses_custom_row_bounds)
...
+                                     get_span_size_by_index{page_indices},
+                                     0,
+                                     thrust::plus<size_t>{});
+    thrust::for_each(
+      rmm::exec_policy_nosync(_stream),
+      iter,
+      iter + total_pages,
+      copy_subpass_page{
+        pass.pages, subpass.page_buf, subpass.page_src_index, dst_offsets, page_indices});
+    subpass.pages = subpass.page_buf;
```
[the remark this answers is not in the document]
Comment (nvdbaranec) [reply cpp/src/io/parquet/reader_impl_chunking.cu:1412]: No. `subpass.pages` is a `hostdevice_span` now.  It wraps either the original vector of pages (if there's no chunking) or the new subset of pages (allocated only if chunking occurs).
https://github.com/rapidsai/cudf/blob/bc4974a9b2f1750c36d2492f9a8843caad8e89a5/cpp/src/io/parquet/reader_impl_chunking.hpp#L72

File: cpp/src/io/parquet/reader_impl_chunking.cu (hunk 5 above)
Comment (nvdbaranec) [reply cpp/src/io/parquet/reader_impl_chunking.cu:1390]: No. See above.

**Review by ttnghia — APPROVED** [review]

**Pull request merged**